lib/PipelineC/Tracing/Runner.cpp

```diff
@@ -0,0 +1,429 @@
+/// \file Runner.cpp
+/// \brief Implements the logic to run a trace file, this boils down to the
+///        Trace.run function that will re-execute the commands of the trace in
+///        order
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <csignal>
+
+#include "llvm/Support/Base64.h"
+#include "llvm/Support/Process.h"
+#include "llvm/Support/Signals.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include "revng/ADT/ConstexprString.h"
+#include "revng/PipelineC/PipelineC.h"
+#include "revng/PipelineC/Tracing/Common.h"
+#include "revng/PipelineC/Tracing/Trace.h"
+#include "revng/Support/Assert.h"
+#include "revng/Support/PathList.h"
+
+#include "Types.h"
+#include "sanitizer/asan_interface.h"
+
+using ArgumentRef = const revng::tracing::Argument &;
+using ArgumentsRef = const llvm::ArrayRef<revng::tracing::Argument>;
+using ReturnRef = const llvm::StringRef;
+
+static void print_assert(bool Condition,
+                         const char *What,
+                         const char *Message,
+                         const char *File,
+                         int Line) {
+  if (Condition)
+    return;
+  dbg << "Warning: Assertion failed in " << File << ':' << Line << ":\n"
+      << What << "\n"
+      << Message << "\n\n";
+}
+
+#define soft_assert(what, message)                               \
+  if (Context.Options.SoftAsserts) {                             \
+    bool Condition = boolcast(what);                             \
+    print_assert(Condition, #what, message, __FILE__, __LINE__); \
+  } else {                                                       \
+    revng_check(what, message);                                  \
+  }
```

Comment:
No macros in non-infrastructure code. Do with functions or inline checks in callers.

Reply:
That would force me to put `__FILE__` and `__LINE__` in each call site, which is not great.
I can abstract the rest in a normal function, but it'd be nice to avoid the above

Reply:
Give on `__FILE__` and `__LINE__`.
There's a lot of other place in the codebase where it'd be nice to have an assertion out of the wrapper, but that's not a justification for spreading macros all over the codebase.
You have the stack trace anyway, so they are not really needed.

Reply:
The entire point of `soft_assert` is that if SoftAsserts are enabled then *the program will not abort*, so accurate file & line are needed (otherwise it becomes quite useless)

Reply:
I don't think, people care about the line of the source file.
What's relevant is the error message and maybe the position of the faulty command in the trace.